main/src/ca/uwaterloo/flix/language/phase/TreeShaker.scala

```diff
@@ -18,29 +18,207 @@ package ca.uwaterloo.flix.language.phase
 
 import ca.uwaterloo.flix.api.Flix
 import ca.uwaterloo.flix.language.CompilationError
-import ca.uwaterloo.flix.language.ast.SimplifiedAst
+import ca.uwaterloo.flix.language.ast.{SimplifiedAst, Symbol}
+import ca.uwaterloo.flix.language.ast.SimplifiedAst.{Expression, Definition}
 import ca.uwaterloo.flix.util.Validation
 import ca.uwaterloo.flix.util.Validation._
 
+import scala.collection.mutable
+
 /**
   * The Tree Shaking phase removes all unused function definitions.
   *
   * A function is considered reachable if it:
   *
   * (a) Appears in the global namespaces, takes zero arguments, and is not marked as synthetic.
   * (b) Appears in a fact or a rule as a filter/transfer function.
-  * (c) Appears in a function which itself is reachable.
+  * (c) Appears in a lattice declaration.
+  * (d) Appears in a function which itself is reachable.
   */
+
 object TreeShaker extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
 
   /**
     * Performs tree shaking on the given AST `root`.
     */
   def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
 
-    // TODO: Implement.
+    /**
+      * A queue of function definitions to be processed recursively.
+      *
+      * For example, if the queue contains the entry:
+      *
+      * -   f
+      *
+      * it means that the function definition f should be considered to determine new reachable functions.
+      */
+    val queue: mutable.Queue[Definition.Constant] = mutable.Queue.empty
+
+    /**
+      * A map used to collect reachable functions (symbols and the corresponding definitions).
+      */
+    val reachableFunctions: mutable.Map[Symbol.DefnSym, Definition.Constant] = mutable.Map.empty
```

Comment:
This could (and should) actually be a set. That would make it clear that you are collecting symbols (and not changing the definitions.)